pkg/cmd/cli/cmd/version.go

```diff
@@ -0,0 +1,168 @@
+	cmd := &cobra.Command{
+		Use:   "version",
+		Short: "Display client and server versions",
+		Long:  versionLong,
+		Run: func(cmd *cobra.Command, args []string) {
+			options.BaseName = fullName
+
+			if err := options.Complete(f, out); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(cmd, err.Error()))
+			}
+
+			if err := options.RunVersion(f); err != nil {
+				kcmdutil.CheckErr(err)
+			}
+		},
+	}
+
+	return cmd
+}
+
+func (o *VersionOptions) Complete(f *clientcmd.Factory, out io.Writer) error {
+	o.Out = out
+	return nil
+}
+
+// RunVersion attempts to display client and server versions for Kubernetes and OpenShift
+func (o VersionOptions) RunVersion(f *clientcmd.Factory) error {
+	fmt.Fprintf(o.Out, "%s %v\n", o.BaseName, version.Get())
+	fmt.Fprintf(o.Out, "kubernetes %v\n", kubeversion.Get())
+	if o.PrintEtcdVersion {
+		fmt.Fprintf(o.Out, "etcd %v\n", etcdversion.Version)
+	}
+
+	if o.PrintClientFeatures {
+		features := []string{}
+		if tokencmd.BasicEnabled() {
+			features = append(features, "Basic-Auth")
+		}
+		if tokencmd.GSSAPIEnabled() {
+			features = append(features, "GSSAPI")
+			features = append(features, "Kerberos") // GSSAPI or SSPI
+			features = append(features, "SPNEGO")   // GSSAPI or SSPI
+		}
+		fmt.Printf("features: %s\n", strings.Join(features, " "))
+	}
+
+	if f == nil {
+		return nil
+	}
+
+	// max amount of time we want to wait for server to respond
+	timeout := 10 * time.Second
+
+	done := make(chan error)
+	oVersion := ""
+	kVersion := ""
+	versionHost := ""
+
+	// start goroutine to fetch openshift / kubernetes server version
+	go func(f *clientcmd.Factory) {
+		defer close(done)
+
+		// confirm config exists before makig request to server
+		var err error
+		o.ClientConfig, err = f.OpenShiftClientConfig.ClientConfig()
+		if err != nil {
+			done <- err
+			return
+		}
+		versionHost = o.ClientConfig.Host
+
+		o.Client, o.KClient, err = f.Clients()
+		if err != nil {
+			done <- err
+			return
+		}
+
+		ocVersionBody, err := o.Client.Get().AbsPath("/version/openshift").Do().Raw()
+		if err != nil && !(kapierrors.IsNotFound(err) || kapierrors.IsUnauthorized(err)) {
+			done <- err
+			return
+		}
+		var ocServerInfo version.Info
+		err = json.Unmarshal(ocVersionBody, &ocServerInfo)
+		if err != nil {
+			done <- err
+			return
+		}
+		oVersion = fmt.Sprintf("%v", ocServerInfo)
+
+		kubeVersionBody, err := o.KClient.Get().AbsPath("/version").Do().Raw()
+		if err != nil && !(kapierrors.IsNotFound(err) || kapierrors.IsUnauthorized(err)) {
+			done <- err
+			return
+		}
+		var kubeServerInfo kubeversion.Info
+		err = json.Unmarshal(kubeVersionBody, &kubeServerInfo)
+		if err != nil {
+			done <- err
+			return
+		}
+		kVersion = fmt.Sprintf("%v", kubeServerInfo)
+
+	}(f)
+
+	select {
+	case err, closed := <-done:
+		if strings.HasSuffix(fmt.Sprintf("%v", err), "connection refused") || clientcmd.IsConfigurationMissing(err) || kclientcmd.IsConfigurationInvalid(err) {
```

Comment:
It turns out that the `IsConfigurationInvalid` function in https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/validation.go#L95 does not take into account config files that are completely missing, just ones with invalid values. Since the error for a missing config file is located in our `clientcmd` package, I went ahead and placed the method `IsConfigurationMissing` there.